lib/Db/MessageMapper.php

```diff
@@ -705,39 +705,32 @@ public function deleteAll(Mailbox $mailbox): void {
 	}
 
 	public function deleteByUid(Mailbox $mailbox, int ...$uids): void {
-		$messageIdQuery = $this->db->getQueryBuilder();
+		$messageIdSubQuery = $this->db->getQueryBuilder();
 		$deleteRecipientsQuery = $this->db->getQueryBuilder();
 		$deleteMessagesQuery = $this->db->getQueryBuilder();
 
-		// Get all message ids query
-		$messageIdQuery->select('id')
+		$messageIdSubQuery->select('id')
 			->from($this->getTableName())
 			->where(
-				$messageIdQuery->expr()->eq('mailbox_id', $messageIdQuery->createNamedParameter($mailbox->getId())),
-				$messageIdQuery->expr()->in('uid', $messageIdQuery->createParameter('uids'))
+				$messageIdSubQuery->expr()->eq('mailbox_id', $deleteRecipientsQuery->createNamedParameter($mailbox->getId())),
+				$messageIdSubQuery->expr()->in('uid', $deleteRecipientsQuery->createParameter('uids'))
 			);
-
 		$deleteRecipientsQuery->delete('mail_recipients')
-			->where($deleteRecipientsQuery->expr()->in('message_id', $deleteRecipientsQuery->createParameter('messageIds')));
+			->where($deleteRecipientsQuery->expr()->in('message_id', $deleteMessagesQuery->createFunction($messageIdSubQuery->getSQL())));
 
 		$deleteMessagesQuery->delete($this->getTableName())
-			->where($deleteMessagesQuery->expr()->in('id', $deleteMessagesQuery->createParameter('messageIds')));
+			->where(
+				$deleteMessagesQuery->expr()->eq('mailbox_id', $deleteMessagesQuery->createNamedParameter($mailbox->getId())),
+				$deleteMessagesQuery->expr()->in('uid', $deleteMessagesQuery->createParameter('uids')),
+			);
 
 		foreach (array_chunk($uids, 1000) as $chunk) {
-			$messageIdQuery->setParameter('uids', $chunk, IQueryBuilder::PARAM_INT_ARRAY);
-			$cursor = $messageIdQuery->executeQuery();
```

Comment:
The original implementation of deleteByUid did use a sub query too: https://github.com/nextcloud/mail/pull/4663. Then https://github.com/nextcloud/mail/pull/4947 added an intermediate fetch. @miaulalala do you remember why this was done?

Reply:
Most likely because I ran into the usual subquery issues (where to put the brackets) and didn't know how to fix it back then.

Reply:
Good. Then this change is safe :)